_posts/2019-08-29-using-tf-from-r/using-tf-from-R.Rmd

```diff
@@ -0,0 +1,440 @@
+```
+
+
+where `r_to_py` is a generic (like `py_to_r` above), `r_to_py_impl` is wrapped by Rcpp and `r_to_py_cpp` is a C++ function that branches on the type of the object -- basically the counterpart of the C++ `r_to_py`.
+
+Now in addition to that general process, there is more going on when we call an R function from Python. As Python doesn't "speak" R, we need to wrap the R function in CPython - basically, we are extending Python here! How to do this is described in the official [Extending Python Guide](https://docs.python.org/3/extending/index.html#extending-index).
+
+In official terms, what `reticulate` does it _embed_ and _extend_ Python.
+Embed, because it lets you use Python from inside R. Extend, because to enable Python to call back into R it needs to wrap R functions in C, so Python can understand them.
+
+As part of the former, the desired Python is loaded (`Py_Initialize()`); as part of the latter, two functions are defined in a new module named `rpycall`, that will be loaded when Python itself is loaded.
+
+```{Rcpp}
+PyImport_AppendInittab("rpycall", &initializeRPYCall);
+```
+
+These methods are `call_r_function`, used by default, and `call_python_function_on_main_thread`, used in cases where we need to make sure the R function is called on the main thread: 
+
+```{Rcpp}
+PyMethodDef RPYCallMethods[] = {
+  { "call_r_function", (PyCFunction)call_r_function,
+    METH_VARARGS | METH_KEYWORDS, "Call an R function" },
+  { "call_python_function_on_main_thread", (PyCFunction)call_python_function_on_main_thread,
+    METH_VARARGS | METH_KEYWORDS, "Call a Python function on the main thread" },
+  { NULL, NULL, 0, NULL }
+};
+```
+
+`call_python_function_on_main_thread` is especially interesting. The R runtime is single-threaded; while the CPython implementation of Python effectively is as well, due to the [Global Interpreter Lock](https://en.wikipedia.org/wiki/Global_interpreter_lock), this is not automatically the case when other implementations are used, or C is used directly. So `call_python_function_on_main_thread` makes sure that unless we can execute on the main thread, we wait.
+
+
+That's it for our three "spotlights on `reticulate`". 
+
+## Wrapup
+
+It goes without saying that there's a lot about `reticulate` we didn't cover in this article, such as memory management, initialization, or specifics of data conversion. Nonetheless, we hope we were able to shed a bit of light on the _magic_ involved in calling TensorFlow from R.
+
+R is a concise and elegant language, but to a high degree its power comes from its packages. Hearing someone say such a thing, we naturally think of the `tidyverse`; but there's also the packages that allow you to call into, and interact with, the outside world, such as deep learning frameworks or distributed processing engines. In this post, it was a special pleasure to focus on a central building block that makes much of this possible: `reticulate`. 
```

Comment:
yeah I see what you mean... strangely, I imagined that _not_ coming from a tidyverse team member it might work ... but that's probably a bit naive ...will remove